support/k8s/k8s-workload-registrar/mode-crd/controllers/node_controller.go

```diff
@@ -0,0 +1,148 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/spiffe/spire/proto/spire/api/registration"
+	"github.com/spiffe/spire/proto/spire/common"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// NodeReconcilerConfig holds the config passed in when creating the reconciler
+type NodeReconcilerConfig struct {
+	Ctx         context.Context
+	Log         logrus.FieldLogger
+	R           registration.RegistrationClient
+	Mgr         ctrl.Manager
+	TrustDomain string
+	Cluster     string
+}
+
+// NodeReconciler holds the runtime configuration and state of this controller
+type NodeReconciler struct {
+	client.Client
+	c NodeReconcilerConfig
+}
+
+// NewNodeReconciler creates a new NodeReconciler object
+func AddNodeReconciler(config NodeReconcilerConfig) error {
+	r := &NodeReconciler{
+		Client: config.Mgr.GetClient(),
+		c:      config,
+	}
+
+	err := ctrl.NewControllerManagedBy(config.Mgr).
+		For(&corev1.Node{}).
+		Complete(r)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile steps through the endpoints for each service and adds the name of the service as
+// a DNS name to the SPIFFE ID CRD
+func (n *NodeReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
```

Comment:
This controller is going to leak entries for nodes deleted while the controller is unavailable. The entire reason to use the custom resource behind the pod controller is to avoid this occurring for pods, the mechanism should exist for Nodes too. That probably means allowing arbitrary parent IDs on SpiffeId, and this controller creating SpiffeId resources in its own namespace for the nodes. (it's permissable for a namespaced resource to be owned by a cluster scoped resource, so this should work.)

Reply:
Good catch, thank you.

Reply:
This is now fixed